### DISCUSSION TITLE
Debian 10 (32-bits) and 11 (64-bits): Does Pygments work with Text Widgets, GUI software, e.g. tkinter & PySimpleGui?

### DISCUSSION THREAD

**Discussion by SymbReprUnlim** [pygments/pygments#2439] [Q&A]
Does Pygments work with Text Widgets GUI software, like tkinter and PySimpleGui? If so, are there any Python demo or sample programs available to use as developmental examples or templates? This is specifically related to Debian Buster 10 32-bit software, but also Debian Bullseye 11 64-bit software. Given that Pygments installs using pip, it obviously works with Linux and therefore Debian and likely Ubuntu.

**Comment by birkenfeld** [comment]
Pygments will be available in these Debian distributions as a system package, if you don't need the most recent version.

Regarding GUIs, I am not familiar with tkinter and PySimpleGui. There are different ways to do it, either they can e.g. directly display HTML from the HTML formatter, or you write your own formatter that inserts the tokens into a widget.

That said, Pygments is not designed for interactive use cases, since the whole text is lexed and highlighted in one go. Therefore, to highlight code in a user-editable text field, you'll have to rehighlight after each change.

**Reply by SymbReprUnlim** [reply]
Thank you for clarifying. Both tkinter(tk) and PySimpleGui(psg) have widget and form refreshes built into them, as essentially endless while or for loops in Python programs, exits occurring when the user left/right clicks some widget or an external interrupt occurs or on/after some discrete time interval. So widget and form or GUI interfaces can refresh at timed intervals automatically or upon certain events. If Pygments and PySimpleGui were integrated to work together, it would open up a broad array of different applications, such as teaching English or other languages interactively with prepositional phrases color coded differently than sentence nouns, verbs, etc. It would likely open up personalized language translators, from one language to another. And it would open up applications in color coded computer language editors or integrated development environments (IDEs) which typically have 1) an editor with program statements color coding, 2) debugging facilities with color coded errors and 3) a colored console, report, form or document output with distinct color coding on it as well. Language translator apps could highlight either sentence by sentence syntax similarities between the two languages OR phrase by phrase similarities or use a traditional language parser syntactical structuring.

Rehighlighting wouldn't likely be problematic. Integrating Pygments with say PSG (more recently developed GUI software than tk) would likely take some coordination and or collaboration. As an initial observation, Pygments and PySimpleGui have similar Operating Systems covered and perhaps other synergies and or syncs in common. Ref. https://www.pysimplegui.org/en/latest/

Each Text display widget and or user Input Text widget, for instance in PSG & tk have had, up to now, a single background color applicable and a distinct foreground or text font color as well, again as the software is currently designed. That's a limitation PSG faces. The limitation Pygments seems to face is user interactivity, varying text colorizing and relexification live, over time, and in real-time. Which again, points to apparent complementary synergies between lexicography, color coding and GUI programs.

Thank you for the open discussion and open-minded frame of reference/approach! Pygments is definitely pushing back - advancing - the lines and limits against what has previously been extremely difficult to program in - lexicography (or more specifically "syntax parsing") combined with color coding.
